src/VirtoCommerce.Platform.Web/wwwroot/js/app/settings/blades/setting-dictionary.js

```diff
@@ -109,32 +129,79 @@ angular.module('platformWebApp').controller('platformWebApp.settingDictionaryCon
         $scope.$watch('blade.parentBlade.currentEntities', function (data) {
             if (data) {
                 var allEntities = _.flatten(_.map(data, _.values));
-                initializeBlade(_.findWhere(allEntities, { name: blade.currentEntityId }));
+                var origEntities = _.flatten(_.map(blade.parentBlade.origEntity, _.values));
+                initializeBlade(_.findWhere(allEntities, { name: blade.currentEntityId }), _.findWhere(origEntities, { name: blade.currentEntityId }));
             }
         });
     }
 
     $scope.checkAll = function () {
         angular.forEach(currentEntities, function (item) {
-            item._selected = blade.selectedAll;
+            // use a field with $dollar-sign prefix to hide such field from angular.equals to avoid "miss-dirtying" issues
+            item.$selected = blade.selectedAll;
         });
     };
 
+    $scope.applyOrder = function () {
+        // Order both: current and source arrays to avoid issues with angular.equals (like "miss-dirtying")
+        orderBy(currentEntities);
+        orderBy(blade.origEntity.allowedValues);
+    };
+
+    $scope.applyValue = function () {
+        // Check the value has no duplicates
+        $scope.error = !$scope.validateDictValue($scope.editValue.value);
+        if (!$scope.error) {
+            $scope.selectedItem.value = $scope.editValue.value;
+            $scope.selectItem(null);
+            $scope.applyOrder();
+        }
+    };
+
+    $scope.inputKeyDown = function ($event) {
+        if ($event.keyCode === 13) {
+            // Apply value on hit Enter
+            $scope.applyValue();
+        };
+        if ($event.keyCode === 27) {
+            // Decline value on hit Esc
+            $scope.selectItem(null);
+        };
+    };
```

Comment:
1. can move to inline html?
2. can use ng-keyup?

Reply:
1. I tried before, but looks bad.
2. OK